frontend/lib/resolution/try-catch-analysis.cpp

```diff
@@ -0,0 +1,376 @@
+          // this is a catchall.
+          const ResolvedExpression& r = rv.byAst(err);
+          auto errType = CompositeType::getErrorType(context);
+          auto dec = ClassTypeDecorator(ClassTypeDecorator::MANAGED_NONNIL);
+          auto manager = AnyOwnedType::get(context);
+          auto ret = ClassType::get(context, errType->basicClassType(),
+                                    manager, dec);
+          auto qt = QualifiedType(QualifiedType::VAR,
+                                  ret->withDecorator(context, dec));
+          if (r.type() == qt) {
+            catchAll = true;
+          }
+        } else {
+          // this is a catchall since it has no error associated with it
+          catchAll = true;
+        }
+        return catchAll;
+    }
+
+    // Returns true if the catches don't cover all of the cases.
+     bool catchesNotExhaustive(const Try* node, RV& rv) {
+      if (node->isTryBang()) {
+        return false;
+      }
+
+      bool hasCatchAll = false;
+      for (auto handler : node->handlers()) {
+        if (isCatchAll(handler, rv)) {
+          hasCatchAll = true;
+          break;
+        }
+      }
+
+      return !hasCatchAll;
+    }
+
+    void enterAst(const AstNode* ast, bool doesThrow) {
+        TryCatchState tca;
+        tca.ast = ast;
+        tca.tryDepth = tryStack.size();
+        tca.fnDoesThrow = doesThrow;
+        tca.taskFunctionDepth = 0;
+        tryStack.push_back(tca);
+    }
+
+    void exitAst() {
+      tryStack.pop_back();
+    }
+
+    void process(const uast::AstNode* symbol,
+                 ResolutionResultByPostorderID& byPostorder);
+
+    bool enter(const Try* ast, RV& rv);
+    void exit(const Try* ast, RV& rv);
+
+    bool enter(const FnCall* fn, RV& rv);
+    void exit(const FnCall* fn, RV& rv);
+
+    // add pairs of methods for Comment, Module, TypeDecl
+
+    bool enter(const Comment* ast, RV& rv);
+    void exit(const Comment* ast, RV& rv);
+
+    bool enter(const Module* ast, RV& rv);
+    void exit(const Module* ast, RV& rv);
+
+    bool enter(const TypeDecl* ast, RV& rv);
+    void exit(const TypeDecl* ast, RV& rv);
+
+    bool enter(const Function* fn, RV& rv);
+    void exit(const Function* fn, RV& rv);
+
+    bool enter(const Throw* ast, RV& rv);
+    void exit(const Throw* ast, RV& rv);
+
+    bool enter(const AstNode* ast, RV& rv);
+    void exit(const AstNode* ast, RV& rv);
+  };
+
+
+  bool TryCatchAnalyzer::enter(const Try* ast, RV& rv) {
+
+    tryDepth++;
+    enterAst(ast, false);
+
+    return true;
+  }
+
+  void TryCatchAnalyzer::exit(const Try* ast, RV& rv) {
+    tryDepth--;
+    exitAst();
+    auto parent = tryStack.back();
+
+    checkCatches(ast, rv);
+
+    // is it an exhaustive catch?
+    bool nonExhaustive = catchesNotExhaustive(ast, rv);
+    // check for a catchall, or this is a try! or the parent function throws
+
+    if (tryDepth == 0 && nonExhaustive && !parent.fnDoesThrow) {
```

Comment:
I haven't tested this, but my hypothesis is that this can be broken as follows:

```Chapel
proc f() {
  try {
    proc g() {
        try {
            blaBlaBla();
        } catch e : OnlyASubClass {

        } // no catchall
    }
  } catch {
    // catch-all, we're good
  }
}
```

In this case, the `tryDepth` at the inner `try` is still 1, so there will be
no complaints about g being a non-throwing function. However, `g` doesn't
actually check all possible exceptions (no catchall).

I think you can better express "try depth" as being the number of consecutive
Try (but not Fn) `TryStates` on top of the stack. To rewrite this check, then,
you'd just check if the back of the stack's `ast` node is a `Try`.

Also, what if you have a catch at the top level (not inside the function).
Won't accessing `tryStack.back()` without checking be bad?